pkg/cloud/azure/azure.go

```diff
@@ -0,0 +1,211 @@
+	}, nil
+}
+
+// PrepareSubmarinerClusterEnv prepares submariner cluster environment on Azure
+// The below tasks will be executed
+// 1. create the inbound and outbound firewall rules for submariner, below ports will be opened
+//    - NAT traversal port (by default 4500/UDP)
+//    - 4800/UDP port to encapsulate Pod traffic from worker and master nodes to the Submariner Gateway nodes
+//    - ESP & AH protocols for private-ip to private-ip gateway communications
+// 2. create the inbound and outbound firewall rules to open 8080/TCP, 8081/TCP port to export metrics service from the
+//    Submariner gateway
+func (r *azureProvider) PrepareSubmarinerClusterEnv() error {
+	if err := r.gwDeployer.Deploy(api.GatewayDeployInput{
+		PublicPorts: []api.PortSpec{
+			{Port: r.nattPort, Protocol: "udp"},
+			{Port: uint16(r.nattDiscoveryPort), Protocol: "udp"},
+			{Port: 0, Protocol: "esp"},
+			{Port: 0, Protocol: "ah"},
+		},
+		Gateways: r.gateways,
+	}, r.reporter); err != nil {
+		return err
+	}
+
+	input := api.PrepareForSubmarinerInput{
+		InternalPorts: []api.PortSpec{
+			{Port: constants.SubmarinerRoutePort, Protocol: "udp"},
+			{Port: r.gatewayMetricsPort, Protocol: "tcp"},
+			{Port: r.globalnetMetricsPort, Protocol: "tcp"},
+		},
+	}
+	err := r.cloudPrepare.PrepareForSubmariner(input, r.reporter)
+	if err != nil {
+		return err
+	}
+
+	r.reporter.Success("The Submariner cluster environment has been set up on Azure")
+	return nil
+}
+
+// CleanUpSubmarinerClusterEnv clean up submariner cluster environment on Azure after the SubmarinerConfig was deleted
+// 1. delete any dedicated gateways that were previously deployed.
+// 2. delete the inbound and outbound firewall rules to close submariner ports
+// 3. delete the inbound and outbound firewall rules to close submariner metrics port
+func (r azureProvider) CleanUpSubmarinerClusterEnv() error {
+	err := r.gwDeployer.Cleanup(r.reporter)
+	if err != nil {
+		return err
+	}
+
+	err = r.cloudPrepare.CleanupAfterSubmariner(r.reporter)
+	if err != nil {
+		return err
+	}
+
+	r.reporter.Success("The Submariner cluster environment has been cleaned up on Azure")
+	return nil
+}
+
+func initializeFromAuthFile(kubeClient kubernetes.Interface, secretNamespace, secretName string) (string, error) {
+	credentialsSecret, err := kubeClient.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	servicePrincipalJSON, ok := credentialsSecret.Data[servicePrincipalJSON]
+	var authInfo struct {
+		ClientId       string
+		ClientSecret   string
+		SubscriptionId string
+		TenantId       string
+	}
+
+	if !ok {
+		return "", errors.New("servicePrincipalJSON is not found in the credentials")
+	}
+
+	err = json.Unmarshal(servicePrincipalJSON, &authInfo)
+	if err != nil {
+		return "", errors.Wrap(err, "error unmarshalling servicePrincipalJSON")
+	}
+
+	if err = os.Setenv("AZURE_CLIENT_ID", authInfo.ClientId); err != nil {
```

Comment:
Is it safe to assume that we can update the environment variable inside the container? Are there any alternate way to pass this value? 

Currently we are executing the cloud prepare in the context of spoke, but when we move this in the context of Hub (in future PR) it might lead to issues as the environment variable will be global in the container image.

Reply:
I was not able to find an alternate way to do this. subctl also uses a similar way. If needs to be moved to hub we need to explore possible options then. It is just to create the authorizer, once created we can use the authorizer to authenticate.